### PR TITLE
[microTVM] Project API infrastructure

### PR DESCRIPTION
**Pull request by areusch**
This PR implements the [Project API RFC](https://github.com/apache/tvm-rfcs/pull/8), a plugin-like infrastructure that allows TVM to interact with build tools for third-party platforms. By implementing the Project API interface for a platform, developers can run core TVM workflows such as host-driven inference and AutoTVM without needing to build the full TVM library on the platform. This is initially targeted at microTVM use cases, where platforms don't have the OS facilities nor the flash space to run a full TVM C++ runtime nor a Python interpreter.

More discussion is available in the RFC. Implementations are provided here for Zephyr and for the POSIX subprocess test platform.

NOTE: This PR is based on top of #8072 and will merge following that one.

@mehrdadh @guberti @tqchen @jroesch @tkonolige @csullivan @leandron @u99127 @Mousius @giuseros @gromero @stoa @hogepodge

### CODE REVIEW

**Review by gromero — CHANGES_REQUESTED** [review]
@areusch Thanks for working on this!

I'll review it bit by bit instead of at once.

Please see my first comment inline.

Besides that, some other general comments:

1) A concern from Matt  (https://discuss.tvm.apache.org/t/rfc-tvm-project-api/9449) still resonates with me: is it really necessary to have two entry points for the server? i.e. a .sh and .py entry point? One would have to craft an RPC server which accepts two pipes (read and write) right? I don't see why we would need that, unless you have other plans for the future :)

2) Instead of a single query (`server_info_query()`) returning a series of predetermined values (in `server.py`, `_dispatch_server_info_query()`) would it be better also to have it accepting a single key a returning only its value instead of a series of values? It's that I believe the number of keys return will increase over time, so it will increasingly return probably more and more data that won't be used at all in some scenarios.

3) From a TVMC perspective, for the `run` command, in the Graph executor case, tvmc will need the JSON for the graph, which will be located in `/model/executor-config/graph/graph.json'`. Of course the project dir was passed to the 'run' in the command line, so it can be easily determined by tvmc. Nonetheless that patch will be kind of hardcoded by the client (tvmc in this case, so I'm wondering if it should actually be returned via the Project API, by  `server_info_query()`. Another way would be to query `server_info_query()` for key `model_library_format_path` and then untar it locally to find the `mod.json` file, however that case also sounds like a shortcut to by pass the Project API design. Could you please share your thoughts  on that case, i.e. what's the correct way for a client to have access to the `mod.json`?

4) That said, I'm still struggling to understand why a RPC client-server model is necessary here. The client starts the server as a subprocess and passes to it (to the server) a read and write local pipe. This is a bit odd to me. A RPC that should have a remote part at some point, but that is used as something quite local afaics. Why something simpler with a  `importlib.import_module` for instance won't work here? The RPC implementation works nice, so I just would like to understand the rational behind it.

Thank you.

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (mehrdadh) [inline apps/microtvm/zephyr/template_project/microtvm_api_server.py:307]: `self._subprocess_env` is not defined. Also `SubprocessEnv` is missing in this file. Maybe copy is from zephyr.py to this file.

Comment (mehrdadh) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:307]: `build_dir` is also undefined

File: python/tvm/micro/compiler.py
Comment (mehrdadh) [inline python/tvm/micro/compiler.py:1]: maybe keep this and add warning that it will be removed in the next version?

**Review by mehrdadh — COMMENTED** [review]
overall looks great!
My question is that whether we need to keep the previous compiler setup and give time to community before deleting it.

File: python/tvm/micro/project.py
Comment (areusch) [reply python/tvm/micro/project.py:71]: could you say why? right now this code is just attempting to document the uses of project_api. i think we don't need to expose that here if it's not called. wdyt?

File: python/tvm/micro/compiler.py
Comment (areusch) [reply python/tvm/micro/compiler.py:1]: so i was thinking about this, one thing is that i had to make breaking changes to the `apps/microtvm/zephyr` code. i don't mind leaving in the old `tvm.micro.Compiler` stuff (though I'd need to adapt more Transport layers), but it might be confusing to have two different styles of Zephyr template projects. we haven't done a TVM release with this code in there, so i wonder if it's necessary to keep the Compiler stuff in there. wdyt?

File: python/tvm/micro/project.py
Comment (gromero) [reply python/tvm/micro/project.py:72]: @areusch I thought of this code as a convenient wrapper around  `TemplateProject` class and, ultimately around project_api.client module. Hence in the use case where a program wants to simply query some info only about the template project (without generating one) it would use the `server_info_query()` method. For instance, with the most recent code there is now a new `options` '`project_type` (which can be currently `host_driven` or `aot_driven`. TVMC create-project command would need to accept an argument specifying which one it can take. Since Client API detects it automatically based on what's in `src/`, it would be matter of querying the `project_type`s  available by using `TemplateProject.server_info_query` to show the user the options available - we would need also create an associated wrapper like `tvmc.micro.project.template_project` On a second run the user woulds already know the right option and pass to TVMC create-project command, which by its turn would that time call `tvm.micro.project.create_project`.

Maybe I've misunderstood some project.py role. Let me know if it's not clear the use case I'm thinking of.

**Review by gromero — CHANGES_REQUESTED** [review]
Hi Andrew, I think
`./aot_demo/boards/qemu_cortex_r5.conf`
`./host_driven/boards/nucleo_l4r5zi.conf`
`./host_driven/boards/qemu_cortex_r5.conf`

were left behind on the move and need to be moved as well under `template_project`?

File: apps/bundle_deploy/crt_config/crt_config.h
Comment (gromero) [inline apps/bundle_deploy/crt_config/crt_config.h:38]: There is still a `crt/` in `template_project` besides `crt_config/`, I believe the former should be deleted?

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (gromero) [inline apps/microtvm/zephyr/template_project/microtvm_api_server.py:397]: I think it's ok to add a list of boards here instead using the `-qemu` suffix trick as in the old API. However I just would like to mention that there are other boards on Zephyr in the same situation, like the new mps3 boards, and probably there will be more in the future. Zephyr doesn't have a policy to enforce the `qemu-` prefix afaik, unfortunately. That's why I introduced the `-qemu` suffix, so one could add more boards like `mps2_an521` to microTVM without having to update the list here and just using the `-qemu` suffix`, so it would be a bit more generic way to handle such a kind of boards.

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (gromero) [inline apps/microtvm/zephyr/template_project/microtvm_api_server.py:74]: I think it could be rewritten as:
```
class CMakeCache(collections.abc.Mapping):
```
so adding `abc` without prejudice in order to avoid deprecation warnings?

**Review by gromero — COMMENTED** [review]
@areusch A couple more comments about the lastest patchset push.

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (areusch) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:397]: ah i see. the main reason i went away from it was that the parameter was named `zephyr_board`, so a new user would be confused why they need to add `-qemu` suffix. happy to entertain other parameter names if you have suggestions!

File: python/tvm/micro/project.py
Comment (areusch) [reply python/tvm/micro/project.py:72]: @gromero ah gotcha. yeah so i view this more as an internal interface rather than an external one. as such, i do think that a server_info_query() may have a place here. maybe better for your use case would be a function `get_project_options` which would call `server_info_query` if it hadn't yet been, and then return the `project_options`. the idea here is to document all of TVM's uses of Project API all in once place. since there isn't that use case today in the code, it isn't there yet. however, you should feel free to add such functions to project.py as needed. 

does that sound ok to you?

File: apps/bundle_deploy/crt_config/crt_config.h
Comment (areusch) [reply apps/bundle_deploy/crt_config/crt_config.h:38]: good catch, i agree.

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (areusch) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:74]: done

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (areusch) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py]: resolved these problems!

File: python/tvm/micro/compiler.py
Comment (areusch) [reply python/tvm/micro/compiler.py:1]: @mehrdadh any thoughts here?

**Comment by areusch** [comment]
@mehrdadh @gromero please take another look!

File: python/tvm/micro/project.py
Comment (gromero) [reply python/tvm/micro/project.py:72]: @areusch Sounds good!

**Comment by gromero** [comment]
@areusch Fixes in `https://github.com/apache/tvm/pull/8380/commits/0d0c504a158ac9e03c301b88a5ca1127c57bd06e` are good. I think the only remaining issue is that `prj.conf` generation mechanism currently doesn't account for boards that don't have a FPU, so `CONFIG_FPU=y` (like the `mps2_an521` board). So I think we need to fix it because it's a regression of https://github.com/apache/tvm/pull/8055. Otherwise I think it's pretty much ready to land once CI is happy too :)

File: python/tvm/micro/compiler.py
Comment (mehrdadh) [reply python/tvm/micro/compiler.py:1]: @areusch  I agree that since there's no official release we don't commit to support older stuff. However I recommend to make it more clear in the PR title/description that this PR is removing certain modules so if anyone has a problem could raise their voice.
The other thing is that you could announce it in the microtvm community meeting and get some feedback there. My concern is that people have internal projects depending on these features.

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (mehrdadh) [inline apps/microtvm/zephyr/template_project/microtvm_api_server.py:289]: `aot_demo` condition is missing?
I think for that the stack size should be 2000

**Review by mehrdadh — CHANGES_REQUESTED** [review]
Overall looks great. Couple of things:
- My understanding is that the api_server only supports qemu boards and the information about running `test_zephyr` or `test_zephyr_aot` on physical board is kinda lost since you're removing board.conf files. What is the plan for this?
- These change will break tflite micro tutorial. I think CI should catch that?

I will read more later and put comments since this is a large PR.

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (gromero) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:289]: @mehrdadh I think CI won't catch the `tutorial/micro/micro_tflite.py`. On the other hand I'm ok fixing it on a follow-up PR.

Comment (gromero) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:289]: @mehrdadh the API server does work on physical boards, I've tested it on disco boards, for instance. `test_zephyr` and `test_zephyr_aot` defaults to use qemu as on the old API, but `prj.conf` generator will take care of generating the proper config file for physical boards too.

Comment (gromero) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:289]: @mehrdadh actually, I've tested the previous version, so I'm re-testing it on disco. Hence it's good to test as you said with the 4 additional boards you have on hand.

**Comment by areusch** [comment]
tested offline against nrf5340 and stm32f746xx_disco; waiting for CI to be fixed to retrigger. @gromero @mehrdadh can you take a look when you have a minute?

**Review by guberti — APPROVED** [review]

**Review by mehrdadh — APPROVED** [review]
There are few follow ups that are mentioned here: https://github.com/apache/tvm/issues/8595

Otherwise LGTM!
Thanks @areusch !

**Comment by gromero** [comment]
@areusch Thanks a lot for addressing all the issues and concerns. LGTM!

**Review by gromero — APPROVED** [review]

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (tmoreau89) [inline apps/microtvm/zephyr/template_project/microtvm_api_server.py:149]: Does the `nrf5340dk_nrf5340_cpuapp` need to be in there?

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (tmoreau89) [inline apps/microtvm/zephyr/template_project/microtvm_api_server.py:281]: you may want to clean up that comment

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (tmoreau89) [inline apps/microtvm/zephyr/template_project/microtvm_api_server.py:310]: Is self-hosted dropped in favor of AOT?

**Review by tmoreau89 — APPROVED** [review]
Thanks @areusch for this heroic PR. I've left a couple comments, but overall LGTM. I'd like @leandron to give the final green light and merge it.

**Review by leandron — APPROVED** [review]
LGTM. Thanks @areusch for the big effort in this **milestone** in the project IMHO.

I discussed with @areusch one small usability point that we well need to aadress later, which is basically that the template project won't be added to Python packages generated from the TVM repo _as is_.

We'll  need to offer some mechanism (and documentation) for a user to point to a Git repo and a path to access the template project to be used, or even maintaining an index of known-good repos and paths we know. This can be done via APIs or indeed via `tvmc` (cc @gromero).

It is probably something that will confues people who are not familiar with the code base and just want to use TVM as a tool.

**Comment by leandron** [comment]
I'm happy for anyone to merge this once the conflicting files are solved and CI is green again. cc @tmoreau89

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (areusch) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:328]: oops, fixed

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (areusch) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py]: good catch!

**Comment by tmoreau89** [comment]
Thank you @areusch , @leandron, @gromero, @guberti, @mehrdadh  the PR was merged!

**Comment by areusch** [comment]
@gromero right now we haven't broadly addressed data dependencies in the wheel. i think we need to at some point build a central e.g. `tvm.utils.get_packaged_resource("...")` function to mimic `tvm.micro.get_standalone_crt_dir` (and use the former internally in the latter). this will allow us to add arbitrary data dependencies to the wheel and refer to their paths later on.

**Pull request merged**
